### PR TITLE
Add spoly(spoint[]) constructor function

### PR DESCRIPTION
**Pull request by vitcpp**

### CODE REVIEW

**Comment by vitcpp** [comment]
After merge of #96, I will update the docs.

File: src/polygon.c [hunk 1]
```diff
@@ -60,7 +60,7 @@ PG_FUNCTION_INFO_V1(spheretrans_poly_inverse);
 PG_FUNCTION_INFO_V1(spherepoly_add_point);
 PG_FUNCTION_INFO_V1(spherepoly_add_points_finalize);
 PG_FUNCTION_INFO_V1(spherepoly_is_convex);
-
+PG_FUNCTION_INFO_V1(spherepoly_from_point_array);
```
Comment (esabol) [inline src/polygon.c:63]: Just wondering why you didn't add this new function to polygon.h? I did in PR #96 and am wondering if I shouldn't have?

File: upgrade_scripts/pg_sphere--1.3.1--1.3.2.sql.in [hunk 2]
```diff
@@ -16,3 +16,12 @@ $$;
...
 DROP OPERATOR IF EXISTS @(spoint, smoc);
+
+CREATE FUNCTION spoly(spoint[])
+  RETURNS spoly
+  AS 'MODULE_PATHNAME', 'spherepoly_from_point_array'
+  LANGUAGE 'c'
+  IMMUTABLE STRICT PARALLEL SAFE;
+
+COMMENT ON FUNCTION spoly(spoint[]) IS
+  'Create spoly from an array of points.';
```
Comment (esabol) [inline upgrade_scripts/pg_sphere--1.3.1--1.3.2.sql.in:27]: Change comment to `creates spoly from an array of points` (no capitalization, no period, and "creates" not "create") in order to match the other comments on the other functions.

**Review by esabol — CHANGES_REQUESTED** [review]

File: sql/poly.sql [hunk 3]
```diff
@@ -82,7 +82,21 @@ SELECT spoly_deg(ARRAY[1.0, 2.0, 3.0, 4.0, 5.0, 6.0]);
...
+
+SELECT spoly(NULL::spoint[]);
+
+SELECT spoly(ARRAY[]::spoint[]);
+
+SELECT spoly(ARRAY[spoint_deg(0, 0)]);
+
+SELECT spoly(ARRAY[spoint_deg(0, 0), spoint_deg(10, 0)]);
+
+SELECT spoly(ARRAY[spoint_deg(0, 0), spoint_deg(10, 0), spoint_deg(10,10)]);
```
Comment (esabol) [inline sql/poly.sql:95]: Nitpick: Change "10,10" to "10, 10".

File: sql/poly.sql [hunk 4]
```diff
@@ -82,7 +82,21 @@ SELECT spoly_deg(ARRAY[1.0, 2.0, 3.0, 4.0, 5.0, 6.0]);
...
+
+SELECT spoly(ARRAY[spoint_deg(0, 0), spoint_deg(10, 0), spoint_deg(10,10), spoint_deg(0, 10)]);
```
Comment (esabol) [inline sql/poly.sql:97]: Nitpick: Change "10,10" to "10, 10".

File: src/polygon.c [hunk 5]
```diff
@@ -951,6 +951,32 @@ spherepoly_deg(PG_FUNCTION_ARGS)
...
+{
+	int			np;
+	ArrayType  *inarr = PG_GETARG_ARRAYTYPE_P(0);
+	SPoint	   *points;
+
+	np = ArrayGetNItems(ARR_NDIM(inarr), ARR_DIMS(inarr));
+
+	if (np < 3)
+	{
+		elog(ERROR, "spoly_deg: invalid number of arguments (must be >= 3)");
```
Comment (esabol) [inline src/polygon.c:965]: Change "spoly_deg" to "spoly" (or "spherepoly_from_point_array"?).

File: src/polygon.c [hunk 6]
```diff
@@ -951,6 +951,32 @@ spherepoly_deg(PG_FUNCTION_ARGS)
...
+		PG_RETURN_NULL();
+	}
+
+	if (ARR_HASNULL(inarr))
+	{
+		elog(ERROR, "spoly_deg: input array is invalid because if has null values");
```
Comment (esabol) [inline src/polygon.c:971]: Change "spoly_deg" to "spoly" (or "spherepoly_from_point_array"?).

File: src/polygon.c (hunk 1 above)
Comment (vitcpp) [reply src/polygon.c:63]: @esabol Function declaration in a header is needed when you use this function in another translation unit (.c file). You just include this header with function declarations to be able to use these functions without any compiler warnings. The spherepoly_from_point_array function is an interface function that is searched (ldsym) and called by postgresql. It is not used in other .c sources. Once, some other functions are defined in .h file, I will add the declaration. 

Furthermore, pgindent requires "extern" modificator for functions declarations. I have some plans to use pgindent to reformat .h files to satisfy the postgresql coding style.

Comment (esabol) [reply src/polygon.c:63]: @vitcpp wrote:
> Function declaration in a header is needed when you use this function in another translation unit (.c file).

Uh, yeah. I've been a professional C developer for over 28 years. I know what `.h` files are used for. 

In _my_ C projects, I put the prototype for _every_ function in its corresponding `.h` file. The only times I don't are when I'm working on an API library and I need to be concerned about which functions are in the public API and which are meant only to be used internally. And even then I'd probably create a separate internal-only `.h` file for the prototypes of internal functions. Anyway, that's just my development philosophy, I guess. Apparently, you have a different philosophy, and that's fine.

Basically, I just want to know if I should remove the prototype for `spherepoly_deg` that I added to `src/polygon.h` in PR #96. It sounds like I should!

File: src/polygon.c (hunk 6 above)
Comment (esabol) [reply src/polygon.c:971]: Also, please change "because if" to "because it" in this error message.

File: src/polygon.c (hunk 1 above)
Comment (vitcpp) [reply src/polygon.c:63]: @esabol I'm sorry, I have no any intentions to teach anyone. I just wanted to clearly describe my position. My bad, a bucket with apples from me! :)

I agree with you to put the function declaration into .h file in pgsphere to follow the pgsphere code style. All other API functions are declared in .h files. I would also propose to add extern, but not in this PR. Pgindent will properly apply formatting to function declarations if it sees the extern keyword.

I will rebase and update my PR in according with your comments after the merge of #96.

Comment (esabol) [reply src/polygon.c:63]: OK, cool.

**Comment by vitcpp** [comment]
I updated the doc with the simple description of spoly and spoly_deg function. I used another approach to describe the functions to demonstrate some other approaches to describe functions. I didn't use funcsynopsis which is not suitable for describing SQL functions (funcsynopsis is for languages like C).

**Review by esabol — APPROVED** [review]
This looks good to me! Thank you so much for this contribution, @vitcpp !
